README.MD
# Enlist a Cloudflare Worker as your Secure Password Scoring and Pwnage Protection API

(Prefer traditional serverless sorcery? [See the AWS Lambda version here](https://github.com/detroitenglish/haveibeenpwned-zxcvbn-lambda-api))

Deploy a private, secure and serverless RESTful endpoint for sanely scoring users' new passwords using Dropbox's [`zxcvbn`](https://github.com/dropbox/zxcvbn) library while (k-)anonymously querying Troy Hunt's [`haveibeenpwned`](https://haveibeenpwned.com/) collection of +10 _billion_ breached accounts.

![API in Action](.github/pwnage.gif?raw=true 'API in Action')

&nbsp;&nbsp;&nbsp;&nbsp;_Example: handling results with [VuetifyJS](https://github.com/vuetifyjs/vuetify)_
<br>
<br>

## Motivation

<a href="https://twitter.com/DetroitEnglish/status/1008276231199055874" target="_blank">People seemed to think this concept was neat</a>, and a tiny RESTful API like this is a perfect use case for [**Cloudflare Workers**](https://developers.cloudflare.com/workers/).

As a bonus, [over 90%](https://twitter.com/troyhunt/status/1011871596603060225) of `pwnedpassword` queries are already cached on Cloudflare's edge, making response times _spooky_ fast 👻

---

## Quick Start

1. Rename `example.cloudflare.env` to `cloudflare.env` and edit the configuration as needed.
2. Install deps with `npm install`
3. Launch 🚀 with `npm run deploy`

### Configuration

The following options are configurable via `cloudflare.env`:

- `ROUTE_PATTERN`: Optionally include a route-matching pattern for your worker. See [`example.cloudflare.env`](./example.cloudflare.env) for details on how patterns are parsed and applied (default: `undefined`)

- `ALLOWED_ORIGIN`: Whitelisted origin for Cross Origin Resource Sharing. If not provided, all origins are allowed (default: `*`)

- `ALLOWED_ORIGIN_PATTERNS`: Comma-seperated list of strings to be converted to `RegExp` patterns for testing the incoming request origin. If the origin matches, it's returned as allowed. If no pattern is matched, CORS falls back to the configured `ALLOWED_ORIGIN` (default: `undefined`)

  - Example: `"(foo|bar|buzz)\.example\.lol,yet\.another\.example\.omg"`

- `CORS_MAXAGE`: Value in seconds for the `Access-Control-Max-Age` CORS header (default: `"300"`)

- `ALWAYS_RETURN_SCORE`: Return the `zxcvbn` score even if the `pwnedpasswords` match value is > 0. See [Response](#Response) for details (default: `undefined`, thereby `false`)

- `CUSTOM_PW_DICT`: Comma-seperated list of words/phrases to be included in the `zxcvbn` strength estimation dictionary. It's a good idea to include e.g. your company or application name here (default: `''`)

- `RETURN_PW_METADATA`: Return the full result of the `zxcvbn` strength estimation as a `metadata` response key. Refer to the [zxcvbn documentation](https://github.com/dropbox/zxcvbn#usage) for details on what that includes (default: `undefined`, thereby `false`)

### Updating

Update configuration à la changes to `cloudflare.env` by re-running `npm run deploy`.

**NOTE**: For reasons unknown, the Cloudflare Dash may continue to show older scripts after a successful upload. Check the `x-worker-last-modified` response header to verify that your latest script is up and running.

### Request

POST user password input to your route as JSON with field `password` like so:

```javascript
// pwned password
{
  "password": "monkey123"
}
```

```javascript
// stronger password
{
  "password": "wonderful waffles"
}
```

### Response

Our little worker-bro will reply with an appropriate status code, and JSON body with `ok` indicating successful scoring and range search, a strength estimation `score` of 0 through 4 per `zxcvbn`, and `pwned` matches, indicating the number times the input appears in the `haveibeenpwned` database.

```javascript
// pwned password 'monkey123'
{
    "ok": true,
    "score": 0,
    "pwned": 56491
}
```

```javascript
// stronger password 'wonderful waffles'
{
    "ok": true,
    "score": 3,
    "pwned": 0
}
```

By default, if `pwned` is greater than 0, then `score` will **always** be 0. You can override this behavior by settings `"ALWAYS_RETURN_SCORE"` to `true` in `cloudflare.env`

If `RETURN_PW_METADATA` is truthy, responses will also include a `metadata` key with the complete `zxcvbn` [strength estimation result object](https://github.com/dropbox/zxcvbn#usage).

Each response will also contain a `x-worker-last-modified` header with a timestamp indicating when the script was built and deployed.

#### Errors

Failure will return JSON to inform you that something's not `ok` and a `message` as to why.

```json
{
  "ok": false,
  "message": "It went kaput 💩"
}
```

### Good to Know

Send a `GET` request to act as a little health-check. Response `204` means you're good to go. Useful for testing CORS configuration 👍

## Because Software

### Disclaimer

I am not affiliated with Cloudflare, Troy Hunt, Dropbox, haveibeenpwned, good software development in general, or any combination thereof.

Handling user passwords is no laughing matter, so handle them with care and respect.

Just like your own users, assume that I have no idea what I'm doing. This part is important, because I have no idea what I'm doing.

**REVIEW THE SOURCE**, and use at your own risk 🙈

### License

MIT


CHANGELOG.MD
# Change Log

## 2.0.0

### BREAKING

- Require Node >= 12

#### Minor Updates

- Add Changelog


.eslintrc.js
module.exports = {
  root: true,
  env: {
    node: true,
    es2021: true,
    browser: true,
  },
  parser: `@babel/eslint-parser`,
  extends: [`eslint:recommended`, `plugin:prettier/recommended`],
  plugins: [`babel`, `prettier`],
  parserOptions: {
    ecmaVersion: 2021,
    sourceType: `module`,
    allowImportExportEverywhere: true,
    babelOptions: {
      configFile: `${__dirname}/.babelrc.js`,
    },
  },
  rules: {
    'no-console': process.env.NODE_ENV === `production` ? `error` : `off`,
    'no-debugger': process.env.NODE_ENV === `production` ? `error` : `off`,
    'require-atomic-updates': 0,
    'babel/no-invalid-this': 0,
    'babel/no-unused-expressions': 0,
    'babel/valid-typeof': 1,
    'no-empty': `off`,
    'no-var': `error`,
    'prefer-template': `error`,
    quotes: [`warn`, `backtick`],
    eqeqeq: `error`,
    strict: `error`,
    'require-await': `error`,
    'prettier/prettier': [
      `warn`,
      {},
      {
        usePrettierrc: true,
      },
    ],
  },
}


webpack.config.js
require(`dotenv`).config({ path: `${__dirname}/cloudflare.env` })

const webpack = require(`webpack`)
const CleanWebpackPlugin = require(`clean-webpack-plugin`)
const CloudflareWorkerPlugin = require(`cloudflare-worker-webpack-plugin`)

const {
  ALLOWED_ORIGIN = `*`,
  ALLOWED_ORIGIN_PATTERNS = ``,
  ALWAYS_RETURN_SCORE = false,
  CORS_MAXAGE = 300,
  RETURN_PW_METADATA = false,
  CUSTOM_PW_DICT = ``,
  ROUTE_PATTERN,
  AUTH_EMAIL,
  AUTH_KEY,
  ZONE_ID,
  SITE_NAME,
} = process.env

if (ALLOWED_ORIGIN_PATTERNS.length) {
  const patterns = ALLOWED_ORIGIN_PATTERNS.split(`,`).map(p =>
    RegExp(p).toString()
  )
  console.info(`Allowed origin patterns: ${patterns.join(`, `)}`)
}

module.exports = {
  entry: `${__dirname}/src/index.js`,
  output: {
    path: `${__dirname}/dist`,
    filename: `index.js`,
  },
  target: `webworker`,
  mode: `production`,
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: `babel-loader`,
        exclude: [/frequency_lists/],
      },
    ],
  },
  performance: {
    hints: false,
  },
  node: false,
  optimization: {
    minimize: false,
  },
  plugins: [
    new CleanWebpackPlugin(`dist/*`, {
      root: __dirname,
      verbose: false,
    }),

    new webpack.DefinePlugin({
      ALLOWED_ORIGIN: JSON.stringify(ALLOWED_ORIGIN),
      ALLOWED_ORIGIN_PATTERNS: JSON.stringify(ALLOWED_ORIGIN_PATTERNS),
      ALWAYS_RETURN_SCORE: JSON.stringify(!!ALWAYS_RETURN_SCORE),
      CORS_MAXAGE: JSON.stringify(CORS_MAXAGE),
      RETURN_PW_METADATA: JSON.stringify(!!RETURN_PW_METADATA),
      CUSTOM_PW_DICT: JSON.stringify(CUSTOM_PW_DICT),
      LAST_MODIFIED: JSON.stringify(new Date().toJSON()),
    }),

    new CloudflareWorkerPlugin(AUTH_EMAIL, AUTH_KEY, {
      enabled: !process.env.NO_UPLOAD,
      zone: ZONE_ID,
      site: SITE_NAME,
      pattern: ROUTE_PATTERN,
      clearRoutes: false,
      skipWorkerUpload: false,
      verbose: true,
      colors: true,
      emoji: true,
    }),
  ],
}


src/index.js
/* eslint-env worker, browser */
/* globals ALLOWED_ORIGIN, ALLOWED_ORIGIN_PATTERNS, ALWAYS_RETURN_SCORE, CORS_MAXAGE, LAST_MODIFIED, CUSTOM_PW_DICT, RETURN_PW_METADATA  */

// this +400kb beast of a library is why we'll let webpack build our worker script
const zxcvbn = require(`zxcvbn`)

// Release the hounds!
addEventListener(`fetch`, event => {
  event.respondWith(judgePassword(event.request))
})

// Rain down the judgement...
async function judgePassword(request) {
  // Ignore any other request methods
  if (!/POST|GET|OPTIONS/.test(request.method)) return await fetch(request)

  let password, metadata

  // set default origin for CORS
  let allowOrigin = ALLOWED_ORIGIN

  let customDict = []

  // origin of incoming request
  const origin = request.headers.get(`origin`)

  if (ALLOWED_ORIGIN_PATTERNS.length) {
    // create array of regex patterns to test origin
    const originPatterns = ALLOWED_ORIGIN_PATTERNS.split(`,`).map(pattern =>
      RegExp(pattern)
    )
    // test request origin against allowed patters
    if (originPatterns.some(pattern => pattern.test(origin))) {
      // if there's a match, return request origin as allowed
      // if not, allowed origin falls back to ALLOWED_ORIGIN (or default '*')
      allowOrigin = origin
    }
  }

  // Add user-defined entries to zxcvbn terms dictionary
  if (CUSTOM_PW_DICT.length) {
    customDict.push(...CUSTOM_PW_DICT.split(`,`))
  }

  // Return failure indication and error message if shit gets weird somehow
  function bail(err = new Error(`Something done goofed 💩`), status = 500) {
    return new Response(
      JSON.stringify({
        ok: false,
        message: err.message,
      }),
      {
        headers: theHeaders(true),
        status,
      }
    )
  }

  // Goofy looking utility for our response headers
  const theHeaders = includeContentType => ({
    'access-control-allow-origin': allowOrigin,
    'access-control-expose-headers': `content-type`,
    'access-control-allow-headers': `content-type`,
    'access-control-max-age': isNumber(CORS_MAXAGE) ? +CORS_MAXAGE : 300,
    'x-worker-last-modified': LAST_MODIFIED,
    'content-type': includeContentType ? `application/json` : void 0,
  })

  // Unless it's a POST, we're done here...
  if (request.method !== `POST`) {
    return new Response(void 0, {
      headers: theHeaders(),
      status: 204,
    })
  }

  try {
    password = await request.json().then(body => body.password)
    if (!password) {
      throw new Error(`JSON body must include the field 'password'`)
    }
    if (!password.length) {
      throw new Error(`Input must include 1 or more characters`)
    }
  } catch (err) {
    return bail(err, 400)
  }

  // run strength estimation and range query in parallel because why not
  let results = await Promise.all([
    zxcvbn(password, customDict),
    checkForPwnage(password),
  ]).catch(err => err)

  // Array or gtfo!
  if (!Array.isArray(results)) {
    return bail(results)
  }

  let [strength, pwned] = results

  let { score } = strength

  // Return full zxcvbn result if so configured
  if (RETURN_PW_METADATA) metadata = strength

  // not sure how this could even happen, but I don't trust me, so we'll validate
  if (!isNumber(score) || !isNumber(pwned)) {
    return bail()
  }

  // always set score to zero if a password is pwned unless configured otherwise
  if (!ALWAYS_RETURN_SCORE && pwned) {
    score = 0
  }

  // The worker has spoken:
  return new Response(JSON.stringify({ ok: true, score, pwned, metadata }), {
    headers: theHeaders(true),
  })
}

// Search haveibeenpwned by range using k-anonymity model. Docs for that here:
//    https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange
async function checkForPwnage(password) {
  // hash the input
  const hash = await sha1(password)

  // snatch the first 5 characters
  const prefix = hash.slice(0, 5)

  // remaining hash characters
  const suffix = hash.slice(5)

  // range-search haveibeenpwned with the prefix
  const range = await fetch(
    `https://api.pwnedpasswords.com/range/${prefix}`
  ).then(data => data.text())

  // if the suffix nowhere in response, there's no pwnage
  if (!range.includes(suffix)) {
    return 0
  }

  // split the API results into an array and search for the suffix
  const match = range.split(`\r\n`).find(r => r.includes(suffix))

  // return number of pwned password matches
  return +match.split(`:`).pop()
}

function isNumber(val) {
  return Number.isSafeInteger(+val)
}

async function sha1(str) {
  const buffer = new TextEncoder(`utf-8`).encode(str)
  return await crypto.subtle.digest(`SHA-1`, buffer).then(hexify)
}

function hexify(buffer) {
  let hexCodes = ``
  const view = new DataView(buffer)
  for (let i = 0; i < view.byteLength; i += 4) {
    hexCodes += view.getUint32(i).toString(16).padStart(8, `0`)
  }
  return hexCodes.toUpperCase()
}
